README.md
# java-case

A simple utility library for selecting different forms depending on the Java major version. In case the Clojure needs
to run on multiple JDK platforms, this can make things more concise.

## Releases and Dependency Information

Latest stable release: 1.1

CLI/deps.edn dependency information:

```clojure
org.clojars.jeff_evans/java-case {:mvn/version "1.1"}
```
Leiningen dependency information:

```clojure
[jeff_evans/java-case "1.1"]
```

This library is not yet deployed to Maven central.

## Usage Examples

```clojure
; assume java.security.KeyStore has been imported
(defn file->keystore
  "Initializes a Java KeyStore from the given filename and password."
  [filename password]
  ;; Java 9 adds a getInstance override to initialize directly
  ;; in older versions, it takes multiple steps
  (java-case/java-case
    "9+" (KeyStore/getInstance (io/file filename) password)
    (doto (KeyStore/getInstance (KeyStore/getDefaultType))
      (.load (io/input-stream (str "file:/" filename)) password))))

(defn instant->local-time
  "Converts the given instant to a LocalTime."
  [instant]
  ;; ofInstant came to LocalTime only JDK 9
  (java-case/java-case
    "9+" (LocalTime/ofInstant instant (ZoneId/of "UTC"))
    (let [ldt (LocalDateTime/ofInstant instant (ZoneId/of "UTC"))]
      (.toLocalTime ldt))))

```

The version expressions can be Strings ending in `+`, in which case intermediate JDK versions are filled in. The order
doesn't matter.

```clojure
(macroexpand-1 '(java-case/java-case
                  "11+"  "Java 11 through 16"
                  "17"   "Java 17!"
                  "1.8+" "Java 1.8 through 10"
                  1.7    "Java 1.7, for some reason"
                  "20+"  "Java 20 and beyond"))
=>
(clojure.core/case
  (clojure.core/or
    us.jeffevans.java-case/*java-spec-version-override*
    (us.jeffevans.java-case/current-java-spec-version))
  ("11" "12" "13" "14" "15" "16")
  "Java 11 through 16"
  "17"
  "Java 17!"
  ("1.8" "9" "10")
  "Java 1.8 through 10"
  "1.7"
  "Java 1.7, for some reason"
  "20"
  "Java 20 and beyond")
```

## Future Proof

This library attempts to be future proof, to support major Java versions that may be released after its latest release.  Suppose `K`=`J+1`, and you have a clause
for `J+`. At macro expansion time, the Clojure compiler detects the Java version is `K` (via the JVM property). This clause should work as expected. However, I
haven't thought of the best way to comprehensively test this (suggestions welcomed).

In any case, you can always just use the default clause to specify the "latest" form.

## Building and Testing

This project was created with https://github.com/seancorfield/clj-new

Run the project's tests

    $ clojure -T:build test

Run the project's CI pipeline and build a JAR

    $ clojure -T:build ci

This will produce an updated `pom.xml` file with synchronized dependencies inside the `META-INF`
directory inside `target/classes` and the JAR in `target`. You can update the version (and SCM tag)
information in generated `pom.xml` by updating `build.clj`.

Install it locally (requires the `ci` task be run first):

    $ clojure -T:build install

Deploy it to Clojars -- needs `CLOJARS_USERNAME` and `CLOJARS_PASSWORD` environment
variables (requires the `ci` task be run first):

    $ clojure -T:build deploy

## License

Copyright © 2021 Jeff Evans

Distributed under the Eclipse Public License version 1.0.


CHANGELOG.md
# Change Log

## 1.0 - 2021-09-24
- Initial release


test/us/jeffevans/java_case_test.clj
(ns us.jeffevans.java-case-test
  (:require [clojure.test :refer :all]
            [us.jeffevans.java-case :as java-case]))

;; white box tests
(deftest version-detecting-test
  (testing "sanity check"
    (is (= (var-get #'java-case/known-java-spec-versions) (java-case/java-spec-versions))))
  (testing "future proof checks"
    (testing "java.specification.version JVM parameter at runtime is higher than we know about"
      (binding [java-case/*java-spec-version-override* "20"]
        (is (= (vec (concat (var-get #'java-case/known-java-spec-versions) ["18" "19" "20"]))
               (java-case/java-spec-versions)))))
    (testing "one of the input's integer-like range boundaries is higher than we know about"
      (is (= (vec (concat (var-get #'java-case/known-java-spec-versions) (map str (range 18 43))))
             (java-case/java-spec-versions (#'java-case/inputs->range-boundaries ["11+" "24+" "42"])))))))

(deftest input-handling-test
  (doseq [[inputs exp-boundaries exp-ranges] [[["1.0"]
                                               {"1.0" [false 0]}
                                               {0 ["1.0"]}]
                                              [["1.0+"]
                                               {"1.0" [true 0]}
                                               {0 (java-case/java-spec-versions)}]
                                              [["1.8" "1.0+"]
                                               {"1.0" [true 1]
                                                "1.8" [false 0]}
                                               {0 ["1.8"]
                                                1 ["1.0" "1.1" "1.2" "1.3" "1.4" "1.5" "1.6" "1.7"]}]
                                              [["9+" "1.7+" "16"]
                                               {"1.7" [true 1]
                                                "9"   [true 0]
                                                "16"  [false 2]}
                                               {0 ["9" "10" "11" "12" "13" "14" "15"]
                                                1 ["1.7" "1.8"]
                                                2 ["16"]}]]]
    (testing inputs
      (testing "inputs->range-boundaries works correctly"
        (is (= exp-boundaries (#'java-case/inputs->range-boundaries inputs))))
      (testing "inputs->version-ranges works correctly"
        (is (= exp-ranges (#'java-case/inputs->version-ranges inputs)))))))

;; macro tests
(deftest java-case-test
  (testing "java-case macro works as expected"
    (doseq [[expected version] [[1 "1.7"]
                                [1 "1.8"]
                                [2 "1.6"]
                                [3 "11"]
                                [4 "1.5"]]]
      (is (= expected (binding [java-case/*java-spec-version-override* version]
                        (java-case/java-case
                          11     3
                          "1.7+" 1
                          1.6    2
                          4)))))))


src/us/jeffevans/java_case.clj
(ns us.jeffevans.java-case
  (:require [clojure.string :as str]))

(def ^:dynamic *java-spec-version-override* nil)

(defn current-java-spec-version
  "Returns the current Java major version, from the java.specification.version JVM property."
  []
  (System/getProperty "java.specification.version"))

(def ^:private ^:const highest-known-spec-version 17)

(def ^:private ^:const known-java-spec-versions
  (conj ["1.0" "1.1" "1.2" "1.3" "1.4" "1.5" "1.6" "1.7" "1.8" "9" "10" "11" "12" "13" "14" "15" "16"]
        (str highest-known-spec-version)))

(defn- highest-version [versions]
  (apply max (map #(Integer. %) (filter #(re-matches #"\d+" %) versions))))

(defn java-spec-versions
  "Gets all recognized Java specification versions. Starts with a fixed, known vector, and expands it if any of the
  following are higher than the highest known version at the time this library was published
  (`highest-known-spec-version`):

  * the `java.specification.version` JVM property value (if an integer)
  * any integer keys from the optional `range-boundaries` parameter

  This is based on the assumption that all future Java spec versions will be increasing integers, from 9 onwards."
  ([]
   (java-spec-versions nil))
  ([range-boundaries]
   (let [int-range-boundaries (filter #(re-matches #"\d+" %) (keys range-boundaries))
         curr-version         (or *java-spec-version-override* (current-java-spec-version))
         highest-v            (highest-version (conj int-range-boundaries curr-version))]
     (if (and highest-v (> highest-v highest-known-spec-version))
       (->> (inc highest-v)
            (range (inc highest-known-spec-version))
            (map str)
            (concat known-java-spec-versions)
            vec)
       known-java-spec-versions))))

(defn- reduce-inputs-kv [acc ^String input idx]
  (if (str/ends-with? input "+")
    (let [ver (.substring input 0 (dec (count input)))]
      (assoc acc ver [true idx]))
    (assoc acc input [false idx])))

(defn- inputs->range-boundaries
  "From an `inputs` vector (version number specifiers), return a map whose keys are version numbers, and whose
  values are two-element vectors. The first item in each vector indicates whether the range at that key is open,
  and the second is the corresponding index back to `inputs` for that entry."
  [inputs]
  (reduce-kv reduce-inputs-kv {} (into {} (keep-indexed (fn [idx input]
                                                           {(str input) idx}) inputs))))

(defn- version-range-reducer [boundaries {:keys [in-range? current-range all-ranges] :as acc} version]
  (cond (contains? boundaries version)
        (assoc acc :in-range? (-> (get boundaries version)
                                  first)
                   :all-ranges (cond-> all-ranges
                                 current-range (assoc (:current-input-idx acc) current-range))
                   :current-range [version]
                   :current-input-idx (-> (get boundaries version)
                                          last))

        in-range?
        (update acc :current-range conj version)

        :else
        acc))

(defn- inputs->version-ranges
  "From an `inputs` vector (version number specifiers), return a map whose keys are version numbers, and whose
  values are two-element vectors. The first item in each vector indicates whether the range at that key is open,
  and the second is the corresponding index back to `inputs` for that entry."
  [inputs]
  (let [boundaries                                           (inputs->range-boundaries inputs)
        {:keys [all-ranges current-input-idx current-range]} (reduce (partial version-range-reducer boundaries)
                                                                     {:all-ranges {}}
                                                                     (java-spec-versions boundaries))]
    (cond-> all-ranges
      current-range
      (assoc current-input-idx current-range))))

(defn- map-indexed-case-exprs [num-defs ranges idx v]
  (if (and (even? idx) (< idx (dec num-defs)))
    (let [version-idx   (/ idx 2)
          version-range (get ranges version-idx)]
      (if (> (count version-range) 1)
        (apply list version-range)
        (first version-range)))
    v))

(defmacro java-case
  "Creates a `case` expression that tests the current Java spec version. Useful for selecting different behavior for
  different JDK versions. `definitions` should be thought of as the clauses to `clojure.core/case` (i.e. alternating
  `java-version` and `result-expr` expressions). The `test-constant` for `case` will be emitted as the current
  `java.specification.version` (or the `*java-spec-version-override*` dynamic var, if set, for testing purposes).

  Each `java-version` expression is expected to be a numeric value or string matching some JDK version. If it's a
  String ending with `+`, then all intermediate major versions will be filled in until the next higher `java-version`
  expression that was given as input. As with `case`, a single default expression may be provided last, which will
  be used if the current JDK version doesn't match any of the given `java-version` expressions.  The specified order of
  the `java-version` constants does not matter.

  This macro attempts to be future proof by assuming that all Java major versions from 9 onwards are increasing integer
  values. As long as this holds, it should be able to work with future Java versions."
  [& definitions]
  (let [num-defs   (count definitions)
        partitions (inputs->version-ranges (map first (partition 2 definitions)))]
    `(case (or *java-spec-version-override* (current-java-spec-version))
           ~@(map-indexed (partial map-indexed-case-exprs num-defs partitions) definitions))))
